### ISSUE TITLE
domain_delta_integs_elem_id for LinearForm

### ISSUE THREAD

**Issue by adam-sim-dev** [mfem/mfem#3567]
Hi all,

I find that when `AddDomainIntegrator` with a `VectorFEDomainLFIntegrator` to a `LinearForm`, I can not specify the `domain_delta_integs_elem_id`. 

In `LinearForm::AssembleDelta()`, `if (!HaveDeltaLocations()) ` seems always to be `true` (is it a bug? I have not seen any places `domain_delta_integs_elem_id` is updated after searching the code), so `
fes->GetMesh()->FindPoints(centers, domain_delta_integs_elem_id, domain_delta_integs_ip);` will always be called to initialize the array `domain_delta_integs_elem_id`.

However, for `mesh->FindPoints()`, I see the comments `
This method is not 100 percent reliable, i.e. it is not guaranteed to find a point, even if it lies inside a mesh element.`

So I am wondering how can I specify the `domain_delta_integs_elem_id` directly if I know it. This will make the results more accurate.

Best,
Adam

**Comment by adam-sim-dev** [comment]
```
void LinearForm::AddDomainIntegrator(LinearFormIntegrator *lfi,
                                     Array<int> &elem_marker)
   DeltaLFIntegrator *maybe_delta =
      dynamic_cast<DeltaLFIntegrator *>(lfi);
   if (!maybe_delta || !maybe_delta->IsDelta())
   {
      domain_integs.Append(lfi);
   }
   else
   {
      domain_delta_integs.Append(maybe_delta);
   }
   domain_integs_marker.Append(&elem_marker);
```
should be updated to
```
   DeltaLFIntegrator *maybe_delta =
      dynamic_cast<DeltaLFIntegrator *>(lfi);
   if (!maybe_delta || !maybe_delta->IsDelta())
   {
      domain_integs.Append(lfi);
      domain_integs_marker.Append(&elem_marker);
   }
   else
   {
      domain_delta_integs.Append(maybe_delta);
      for (int i=0; i< elem_marker.Size(); i++)
      {
         domain_delta_integs_elem_id.Append(elem_marker[i]);
      }
   }
   
```

**Comment by adam-sim-dev** [comment]
> ```
> void LinearForm::AddDomainIntegrator(LinearFormIntegrator *lfi,
>                                      Array<int> &elem_marker)
>    DeltaLFIntegrator *maybe_delta =
>       dynamic_cast<DeltaLFIntegrator *>(lfi);
>    if (!maybe_delta || !maybe_delta->IsDelta())
>    {
>       domain_integs.Append(lfi);
>    }
>    else
>    {
>       domain_delta_integs.Append(maybe_delta);
>    }
>    domain_integs_marker.Append(&elem_marker);
> ```
> 
> should be updated to
> 
> ```
>    DeltaLFIntegrator *maybe_delta =
>       dynamic_cast<DeltaLFIntegrator *>(lfi);
>    if (!maybe_delta || !maybe_delta->IsDelta())
>    {
>       domain_integs.Append(lfi);
>       domain_integs_marker.Append(&elem_marker);
>    }
>    else
>    {
>       domain_delta_integs.Append(maybe_delta);
>       for (int i=0; i< elem_marker.Size(); i++)
>       {
>          domain_delta_integs_elem_id.Append(elem_marker[i]);
>       }
>    }
>    
> ```

Oh this won't fix because `domain_delta_integs_ip` is not updated yet.

**Comment by adam-sim-dev** [comment]
Any comments?

**Comment by mlstowell** [comment]
Hi, @adam-sim-dev,

Sorry for the delay. I would not say this is a bug. It's more likely we just didn't see a viable alternative to FindPoints so we didn't design an alternate interface.

I see a few complications. First, user-provided element/point pairs would become outdated after mesh refinement (either adaptive or uniform). This is a minor issue because `LinearForm::Update` resets these pairs and so an error will be reported should the user not send updated pairs after refinement. Second, what if FindPoints is desired for some delta locations but not others. This may require a change to the `DeltaCoefficient` so that it stores information about how it is to be updated. In either case there's the matter of the sequence of user-provided points. In general there could be several integrators some of which use standard coefficients while others use delta coefficients. This is not that hard to account for but this type of bookkeeping could produce hard to find bugs.

Another possibility is to add an option to FindPoints indicating that a failure to locate points is not an option (I'm curious to hear the thoughts of @kmittal2 on this). We might be able to modify our FindPoint wrapper so that it adjusts the locations slightly to search for elements nearby until it finds one. This can be complicated in parallel when a location is on or near a processor boundary but we may be able to sort this out.

In short, this may be the start of a longer conversation.

Best wishes,
Mark

**Comment by adam-sim-dev** [comment]
Hi Mark @mlstowell , a bug here means:

> In `LinearForm::AssembleDelta()`, `if (!HaveDeltaLocations()) ` seems always to be `true` (is it a bug? I have not seen any places `domain_delta_integs_elem_id` is updated after searching the code), so ` fes->GetMesh()->FindPoints(centers, domain_delta_integs_elem_id, domain_delta_integs_ip);` will always be called to initialize the array `domain_delta_integs_elem_id`.

**Comment by mlstowell** [comment]
Hi, @adam-sim-dev,

To the best of my knowledge this is not a bug. `LinearForm::AssembleDelta` could be called multiple times. The first time it will use `FindPoints` to locate the element/ip pairs for each `DeltaCoefficient`. Subsequent times it will use the cached element/ip pairs. Currently there is no way for the user to override this behavior but I believe this is by design.

I'm sure we would be open to discussing ways to add user-provided element/ip pairs. I'm just saying that would be adding a feature rather than fixing a bug.

Best wishes,
Mark

**Comment by adam-sim-dev** [comment]
Thanks @mlstowell I will close this issue.
